crates/wasi-common/src/fdpool.rs

```diff
@@ -0,0 +1,109 @@
+use std::collections::HashSet;
+use std::fmt;
+use std::hash::Hash;
+
+pub(crate) trait Fd: Copy + Clone + Hash + PartialEq + Eq + Default {
+    fn next(&self) -> Option<Self>;
+}
+
+pub(crate) struct FdPool<T: Fd> {
+    next_alloc: Option<T>,
+    available: Vec<T>,
+    claimed: HashSet<T>,
+}
+
+impl<T: Fd + fmt::Debug> fmt::Debug for FdPool<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("FdPool")
+            .field("next_alloc", &self.next_alloc)
+            .field("available", &self.available)
+            .field("claimed", &self.claimed)
+            .finish()
+    }
+}
+
+impl<T: Fd> FdPool<T> {
+    pub fn new() -> Self {
+        Self {
+            next_alloc: Some(T::default()),
+            available: Vec::new(),
+            claimed: HashSet::new(),
+        }
+    }
+
+    pub fn allocate(&mut self) -> Option<T> {
+        if let Some(fd) = self.available.pop() {
+            // Since we've had free, unclaimed handle in the pool,
+            // simply claim it and return.
+            self.claimed.insert(fd);
+            return Some(fd);
+        }
+        // There are no free handles available in the pool, so try
+        // allocating an additional one into the pool. If we've
+        // reached our max number of handles, we will fail with None
+        // instead.
+        let fd = self.next_alloc.take()?;
+        // It's OK to not unpack the result of `fd.next()` here which
+        // can fail since we check for `None` in the snippet above.
+        self.next_alloc = fd.next();
+        self.claimed.insert(fd);
+        Some(fd)
+    }
+
+    pub fn deallocate(&mut self, fd: T) -> bool {
+        if self.claimed.remove(&fd) {
+            self.available.push(fd);
+            return true;
+        }
+        false
```

Comment:
Hmm, I'm fine with asserts. However, the reason I've designed `deallocate` return `bool` was to mimic the behaviour of `HashSet::remove` which returns a `bool` to signal if removal worked or not (i.e., if any element was actually removed in the end). In this case, I guess you're right that if we try removing from the `claimed` `HashSet` and it actually fails, then something went horribly wrong as it shouldn't ever happen.

Reply:
Oh, after re-reading your comment again, checking whether `fd` is not greater than `next_alloc` will require an `if` anyway since `next_alloc` is wrapped in an `Option` and it will not always be guaranteed to be `Some` (in particular, when we exceed all possible allocs for instance, e.g., `std::u32::MAX`).

Reply:
Given the above, unless you can think of a cleaner way of doing it, I guess I'd be in favour of the original approach. What do you think?

Reply:
Perhaps we could `assert!` on `self.claimed.remove` instead? This would ensure we `panic!` if the client code ever provides an illegal `fd` value?